source/opt/scalar_replacement_pass.cpp

```diff
@@ -0,0 +1,702 @@
+  GetOrCreateInitialValue(varInst, index, inst);
+  get_def_use_mgr()->AnalyzeInstDefUse(inst);
+  context()->set_instr_block(inst, block);
+
+  replacements->push_back(inst);
+}
+
+uint32_t ScalarReplacementPass::GetOrCreatePointerType(uint32_t id) {
+  auto iter = pointee_to_pointer_.find(id);
+  if (iter != pointee_to_pointer_.end()) return iter->second;
+
+  uint32_t ptrId = 0;
+  for (auto global : context()->types_values()) {
+    if (global.opcode() == SpvOpTypePointer &&
+        global.GetSingleWordInOperand(0u) == SpvStorageClassFunction &&
+        global.GetSingleWordInOperand(1u) == id) {
+      if (!context()->get_feature_mgr()->HasExtension(
+              libspirv::Extension::kSPV_KHR_variable_pointers) ||
+          get_decoration_mgr()->GetDecorationsFor(id, false).empty()) {
+        // If variable pointers is enabled, only reuse a decoration-less
+        // pointer of the correct type
+        ptrId = global.result_id();
+        break;
+      }
+    }
+  }
+
+  if (ptrId != 0) {
+    pointee_to_pointer_[id] = ptrId;
+    return ptrId;
+  }
+
+  ptrId = TakeNextId();
+  context()->AddType(MakeUnique<ir::Instruction>(
+      context(), SpvOpTypePointer, 0, ptrId,
+      std::initializer_list<ir::Operand>{
+          {SPV_OPERAND_TYPE_STORAGE_CLASS, {SpvStorageClassFunction}},
+          {SPV_OPERAND_TYPE_ID, {id}}}));
+  ir::Instruction* ptr = &*--context()->types_values_end();
+  get_def_use_mgr()->AnalyzeInstDefUse(ptr);
+  pointee_to_pointer_[id] = ptrId;
+
+  return ptrId;
+}
+
+uint32_t ScalarReplacementPass::GetIntId() {
+  if (int_id_ != 0) return int_id_;
+
+  uint32_t id = 0;
+  for (auto type : context()->types_values()) {
+    if (type.opcode() == SpvOpTypeInt) {
+      if (GetIntegerLiteral(type.GetInOperand(0u)) == 32 &&
+          (GetIntegerLiteral(type.GetInOperand(1u)) == 0 ||
+           GetIntegerLiteral(type.GetInOperand(1u)) == 1)) {
+        id = type.result_id();
+        break;
+      }
+    }
+  }
+  if (id != 0) return id;
+
+  id = TakeNextId();
+  context()->AddType(MakeUnique<ir::Instruction>(
+      context(), SpvOpTypeInt, 0, id,
+      std::initializer_list<ir::Operand>{
+          {SPV_OPERAND_TYPE_LITERAL_INTEGER, {32}},
+          {SPV_OPERAND_TYPE_LITERAL_INTEGER, {0}}}));
+  ir::Instruction* integer = &*--context()->types_values_end();
+  get_def_use_mgr()->AnalyzeInstDef(integer);
+
+  return id;
+}
+
+void ScalarReplacementPass::GetOrCreateInitialValue(ir::Instruction* source,
+                                                    uint32_t index,
+                                                    ir::Instruction* newVar) {
+  assert(source->opcode() == SpvOpVariable);
+  if (source->NumInOperands() < 2) return;
+
+  uint32_t initId = source->GetSingleWordInOperand(1u);
+  uint32_t storageId = GetStorageType(newVar)->result_id();
+  ir::Instruction* init = get_def_use_mgr()->GetDef(initId);
+  uint32_t newInitId = 0;
+  if (init->opcode() == SpvOpConstantNull) {
```

Comment:
There is code in fold_spec_constant_op_and_composite_pass.cpp that does something similar, but I'm currently in the midst of re-factoring it for use with CCP.  Add a TODO for me to deal with this after the dust settles?

Reply:
ok